### PR TITLE
[embedder] embedder external view adopt DisplayListEmbedderViewSlice

### PR DESCRIPTION
**Pull request by cyanglaz**
Adopt `DisplayListEmbedderViewSlice` in EmbedderExternalView

Fixes https://github.com/flutter/flutter/issues/123349

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/embedder/embedder_external_view_dl_op_spy.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+#include "flutter/display_list/dl_op_receiver.h"
+
+namespace flutter {
+
+//------------------------------------------------------------------------------
+/// Receives to drawing commands of a DisplayListBuilder.
+///
+/// This is used to determine whether anything will be drawn
+/// on the canvas.
+class EmbedderExternalViewDlOpReceiver final : public virtual DlOpReceiver {
```
Comment (flar) [inline shell/platform/embedder/embedder_external_view_dl_op_spy.h:17]: Have this inherit from the various "Ignore" helpers in dl_receiver_utils then you don't have to create all of the empty overrides.

You can even do that for the attributes and then override the ones you want to handle and leave the rest to be ignored by the helper.

File: shell/platform/embedder/embedder_external_view_dl_op_spy.cc [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                                                 const DlColor colors[],
+                                                 int count,
+                                                 DlBlendMode mode,
+                                                 DlImageSampling sampling,
+                                                 const SkRect* cull_rect,
+                                                 bool render_with_attributes) {}
+void EmbedderExternalViewDlOpReceiver::drawDisplayList(
+    const sk_sp<DisplayList> display_list,
+    SkScalar opacity) {
+  FML_DLOG(ERROR) << "drawDisplayList ";
```
Comment (flar) [inline shell/platform/embedder/embedder_external_view_dl_op_spy.cc:147]: For this method, I would do something like:

```
  Something like [auto remembered_transparency_info = current_transparency_info_;]
  Something like [current_transparency_info_ = defaults;]
  display_list->Dispatch(this);
  Something like [current_transparency_info_ = remembered_transparency_info]
```

Basically, play back the DL with the following caveats:
- Your local attributes should be unaffected by the actions of the DL
- The DL should be played back assuming default attributes
- Your local attributes when everything is done should be the same state as before you recursed.

That is why I have the "save/restore" lines for whatever "current_transparency_info" might be.

Comment (flar) [reply shell/platform/embedder/embedder_external_view_dl_op_spy.cc:147]: Another option would be:

```
  EmbedderExternalViewDlOpReceiver sub_receiver;
  display_list->Dispatch(sub_receiver);
  this->did_draw_ = this->did_draw_ || sub_receiver.did_draw_;
```

Comment (flar) [reply shell/platform/embedder/embedder_external_view_dl_op_spy.cc:147]: And, of course, if "did_draw" is already true, there is no need to recurse in this call as it can't add any new info for your eventual reported answer...

File: shell/platform/embedder/embedder_external_view.cc [hunk 3]
```diff
@@ -59,7 +62,10 @@ bool EmbedderExternalView::HasPlatformView() const {
 }
 
 bool EmbedderExternalView::HasEngineRenderedContents() const {
-  return canvas_spy_->DidDrawIntoCanvas();
+  slice_->end_recording();
```
Comment (flar) [inline shell/platform/embedder/embedder_external_view.cc:65]: Just be sure that this is not called before the rendering is actually done...

File: shell/platform/embedder/embedder_external_view.cc [hunk 4]
```diff
@@ -95,11 +101,15 @@ bool EmbedderExternalView::Render(const EmbedderRenderTarget& render_target) {
...
   }
-
-  canvas->setMatrix(surface_transformation_);
-  canvas->clear(SK_ColorTRANSPARENT);
-  canvas->drawPicture(picture);
-  canvas->flush();
+  auto dl_canvas = std::make_unique<DlSkCanvasAdapter>(canvas);
+  int restore_count = dl_canvas->GetSaveCount();
+  dl_canvas->SetTransform(surface_transformation_);
+  dl_canvas->Clear(SK_ColorTRANSPARENT);
```
Comment (flar) [inline shell/platform/embedder/embedder_external_view.cc:107]: Technically, while SK_Color constants are automatically interchangeable with DlColor instances (they both just boil down to a uint32_t), It would be better to use `DlColor::kTransparent()` here to continue the theme of non-skia use while talking to Dl objects.

File: shell/platform/embedder/embedder_external_view.cc [hunk 5]
```diff
@@ -95,11 +101,15 @@ bool EmbedderExternalView::Render(const EmbedderRenderTarget& render_target) {
...
+  slice_->render_into(dl_canvas.get());
+  dl_canvas->RestoreToCount(restore_count);
+
+  // canvas->drawPicture(picture);
+  // canvas->flush();
```
Comment (flar) [inline shell/platform/embedder/embedder_external_view.cc:112]: `dl_canvas.Flush()`

**Comment by cyanglaz** [comment]
@flar This is ready to review!

File: display_list/BUILD.gn [hunk 6]
```diff
@@ -123,6 +126,7 @@ if (enable_unittests) {
       ":display_list_fixtures",
       "//flutter/display_list/testing:display_list_testing",
       "//flutter/testing",
+      "//third_party/skia",
```
Comment (flar) [inline display_list/BUILD.gn:129]: Why is this needed? I'm guessing if the spy stuff gets moved out of here into the embedder source directories then it doesn't need to be added here, but I would have thought that you could implement an OpReceiver without having to add this dependency...?

File: display_list/dl_op_spy.cc [hunk 7]
```diff
@@ -0,0 +1,171 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/display_list/dl_op_spy.h"
+
+namespace flutter {
+
+bool DlOpSpy::DidDraw() {
```
Comment (flar) [inline display_list/dl_op_spy.cc:9]: I've been using lower case for getters which I think is compliant with the style guidelines - `did_draw()`...

File: display_list/dl_op_spy.cc [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+void DlOpSpy::setColor(DlColor color) {
+  if (color.isTransparent()) {
+    dirty_ = false;
+  } else {
+    dirty_ = true;
+  }
+}
+void DlOpSpy::setColorSource(const DlColorSource* source) {
+  const DlColorColorSource* color_source = source->asColor();
```
Comment (flar) [inline display_list/dl_op_spy.cc:21]: What if source is null?

File: display_list/dl_op_spy.cc [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_;
+}
+void DlOpSpy::drawLine(const SkPoint& p0, const SkPoint& p1) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && p0 != p1;
```
Comment (flar) [inline display_list/dl_op_spy.cc:44]: p0 == p1 draws something

File: display_list/dl_op_spy.cc [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+}
+void DlOpSpy::drawRect(const SkRect& rect) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !rect.isEmpty();
```
Comment (flar) [inline display_list/dl_op_spy.cc:50]: I'd have to check, but I think an empty rect will draw something in stroke mode just like a zero length line. Maybe there is a check somewhere that avoids drawing an empty rect even in stroke mode, but I'd have to search to make sure.

File: display_list/dl_op_spy.cc [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !outer.isEmpty() && outer != inner;
+}
+void DlOpSpy::drawPath(const SkPath& path) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !path.isEmpty();
```
Comment (flar) [inline display_list/dl_op_spy.cc:80]: Double check if path.isEmpty really means there is no geometry. If it means the bounds are empty, then we might end up stroking something in stroke mode.

File: display_list/dl_op_spy.cc [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+}
+void DlOpSpy::drawShadow(const SkPath& path,
+                         const DlColor color,
+                         const SkScalar elevation,
+                         bool transparent_occluder,
+                         SkScalar dpr) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = !color.isTransparent() && elevation > 0 && !path.isEmpty();
```
Comment (flar) [inline display_list/dl_op_spy.cc:168]: I'm not sure about elevation == 0. I think it might also depend on transparent_occluder as well.

File: shell/platform/embedder/embedder_external_view.cc [hunk 13]
```diff
@@ -59,7 +58,9 @@ bool EmbedderExternalView::HasPlatformView() const {
 }
 
 bool EmbedderExternalView::HasEngineRenderedContents() const {
-  return canvas_spy_->DidDrawIntoCanvas();
+  TryEndRecording();
+  slice_->dispatch(*dl_op_spy_);
+  return dl_op_spy_->DidDraw() && !slice_->is_empty();
```
Comment (flar) [inline shell/platform/embedder/embedder_external_view.cc:63]: If this method is called more than once it will repeat the op_spy operation. Also, it will re-use the op_spy object which means the second answer might not be correct. I would change this so that it only does the spy operation once and remembers the answer.

Also, you don't need to permanently store an op_spy object, just create one on the fly when you need to compute the answer and remove the dl_op_spy field from the object.

**Review by flar — CHANGES_REQUESTED** [review]

File: display_list/BUILD.gn (hunk 6 above)
Comment (cyanglaz) [reply display_list/BUILD.gn:129]: This is for the unittest, to generate a concrete DlImage for testing.

File: ci/licenses_golden/licenses_flutter [hunk 14]
```diff
@@ -730,6 +730,8 @@ ORIGIN: ../../../flutter/display_list/dl_op_receiver.cc + ../../../flutter/LICEN
 ORIGIN: ../../../flutter/display_list/dl_op_receiver.h + ../../../flutter/LICENSE
 ORIGIN: ../../../flutter/display_list/dl_op_records.cc + ../../../flutter/LICENSE
 ORIGIN: ../../../flutter/display_list/dl_op_records.h + ../../../flutter/LICENSE
+ORIGIN: ../../../flutter/display_list/dl_op_spy.cc + ../../../flutter/LICENSE
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply ci/licenses_golden/licenses_flutter:733]: I initially put it in the platform/embedder/ but saw `canvas_spy` is in /common, so I thought maybe this class should be in a common place. 

Do you have a preference where it should be?

File: display_list/dl_op_spy.cc (hunk 10 above)
Comment (cyanglaz) [reply display_list/dl_op_spy.cc:50]: What does it draw? A single pixel? Do you know if there is an existing test or document for this we can check?

File: display_list/dl_op_spy.cc (hunk 12 above)
Comment (cyanglaz) [reply display_list/dl_op_spy.cc:168]: what does `transparent_occluder` do?

File: display_list/dl_op_spy.h [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+//------------------------------------------------------------------------------
+/// Receives to drawing commands of a DisplayListBuilder.
+///
+/// This is used to determine whether anything will be drawn
+/// on the canvas.
+class DlOpSpy final : public IgnoreAttributeDispatchHelper,
+                      public IgnoreClipDispatchHelper,
+                      public IgnoreTransformDispatchHelper,
+                      public IgnoreDrawDispatchHelper {
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply display_list/dl_op_spy.h:21]: There are still calls in ignoreDraw: save, savelayer, restore.

**Comment by cyanglaz** [comment]
@flar Updated per review comments and offline discussions. PTAL.

**Review by flar — APPROVED** [review]
Minor edits suggested, but the basic concept is good.

**Review by flar — APPROVED** [review]
Reiterating LGTM after review updates...
